chickensoft-games/AutoInject
# feat: [Issue20] Add WeakReference to store reference types in a default provider

**Context**
When storing a reference type variable as a dependency value created in the DefaultProvider, it was causing the assemblies to fail to unload within the Godot Editor every time a build was re-built. This only appears to happen in the case where a DefaultProvider is used, likely because a strong reference is created within the AutoInject framework instead of within the Godot game using the framework. 

Within the AutoInject framework, this reference should be a WeakReference instead of a strong reference to allow the Godot game to manage the reference correctly and garbage collect it when appropriate.

**Considerations**

1. Does adding the weak reference create a performance issue?
2. Does storing value types in an `object` type create a performance issue?

For these considerations, the best I understand is, previously storing value types within a `dynamic` causes them to be boxed and unboxed, which is the same as when storing them in an `object` type. 
As well, the WeakReference acts like a boxing and unboxing as well. I believe this means that there is no performance changes between this solution and the previous one.

**Testing**
To test this, create a simple class with a dependency on a `Godot.Resource` type with the previous version of AutoInject.
Example:
```
[Meta(typeof(IDependent)), GlobalClass, Tool]
public partial class DependentNode: Node {
    #region Dependency Injection
    [Dependency] private Resource someResource => this.DependOn(
        () => new Resource()
    );
    #endregion
}
```
Following these steps you should be able to see the error:

1. Create the above node within the scene tree.
2. Make a change to the code of the above node.
3. This forces you to rebuild the project.
4. When rebuilding, you should encounter the following error and the project will need a reload to work again:
```
modules/mono/mono_gd/gd_mono.cpp:522 - .NET: Failed to unload assemblies. Please check https://github.com/godotengine/godot/issues/78513 for more information. (User)
  modules/mono/csharp_script.cpp:676 - Condition "managed_callable->delegate_handle.value == nullptr" is true. Continuing.
```

Following the same steps, with this branch, you should see no error.
